site/_posts/2017-11-27-ray-0.3-release.markdown

```diff
@@ -0,0 +1,158 @@
+
+        i += config['beta']
+        time.sleep(0.01)
+
+register_trainable('my_func', my_func)
+
+run_experiments({
+    'my_experiment': {
+        'run': 'my_func',
+        'resources': {'cpu': 1, 'gpu': 0},
+        'stop': {'mean_accuracy': 100},
+        'config': {
+            'alpha': grid_search([0.2, 0.4, 0.6]),
+            'beta': grid_search([1, 2]),
+        },
+    }
+})
+```
+
+In-progress results can be visualized live using tools such as Tensorboard and
+rllab's VisKit (or you can read the JSON format logs directly from the driver
+node):
+
+View the [documentation][2] and the [code][9].
+
+### RLlib integration with Ray.tune
+
+You can try out RLlib with Ray.tune with the following example.
+
+```bash
+cd ray/python/ray/rllib
+python train.py -f tuned_examples/cartpole-grid-search-example.yaml
+```
+
+The `tuned_examples` directory also contains pre-tuned hyperparameter
+configurations for common benchmark tasks such as Pong and Humanoid. View the
+[RLlib documentation][3].
+
+### Initial PyTorch support for RLlib
+
+A modern reinforcement learning library should work with multiple deep learning
+frameworks. As a step toward this goal, 0.3 adds support for PyTorch models for
+A3C in RLlib. You can try this out with the following A3C config.
+
+```bash
+cd ray/python/ray/rllib
+./train.py --run=A3C \
+           --env=PongDeterministic-v4 \
+           --config='{"use_pytorch": true, "num_workers": 8, "use_lstm": false, "model": {"grayscale": true, "zero_mean": false, "dim": 80, "channel_major": true}}'
+```
+
+## Distributed Actors
+
+Ray 0.3 adds support for [distributed actors][1], that is, the ability to have
+multiple callers invoke methods on the same actor. The actor’s creator may pass
+the actor handle as an argument to other tasks or to other actor methods. Here’s
+an example in which the driver creates an actor to log messages and passes an
+actor handle to other tasks:
+
+We've already found actor handles useful for implementing [**parameter
+server**][10] and [**streaming MapReduce**][11] applications.
+
+```python
+import ray
+
+ray.init()
+
+@ray.remote
+class Logger(object):
+    def __init__(self):
+        self.logs = []
+    def log(self, log_msg):
+        self.logs.append(log_msg)
+    def read_logs(self):
+        return self.logs
+
+@ray.remote
+def task(logger, task_index):
+    # Do some work.
+    logged = logger.log.remote('Task {} is done'.format(task_index))
+
+# Create an actor and get a handle to it.
+logger = Logger.remote()
+# Pass the actor handle to some tasks.
+tasks = [task.remote(logger, i) for i in range(10)]
+# Wait for the tasks to finish.
+ray.get(tasks)
+# Get the logs from the tasks.
+print(ray.get(logger.read_logs.remote()))
+```
+
+This feature is still under development, and full functionality is coming soon.
+Currently, the feature behaves as follows.
+
+- Execution order follows submission order, per handle. This means that Ray only
```

Comment:
- all actor method invocations are serialized, i.e., all actor methods are reentrant.
- actor method invocation order : actor methods invoked on the same handle are executed in the order of invocation. No ordering guarantees are provided for method calls invoked on different actor handles.